### PR TITLE
Fixes #2823 remove redundant NOT_INITILIZED evens and change event or…

### LINKED ISSUE #2823: Fix events and event order fired by the automation module
When a rule is added the first NOT_INITIALIZED event is redundant.
For example:
osgi> 14:28:16.652 INFO  s.event.RuleStatusInfoEvent[:43] - rule_3 updated: NOT_INITIALIZED
14:28:16.657 INFO  s.event.RuleStatusInfoEvent[:43] - rule_3 updated: IDLE
14:28:16.658 DEBUG o.e.s.c.c.r.AbstractManagedProvider[:62] - Added new element rule_3 to ManagedRuleProvider.
14:28:16.658 INFO  smarthome.event.RuleAddedEvent[:43] - Rule 'rule_3' has been added.

or when the rule can't be resolved, the following events are fired:
osgi> 14:28:16.652 INFO  s.event.RuleStatusInfoEvent[:43] - rule_3 updated: NOT_INITIALIZED
14:28:16.657 INFO  s.event.RuleStatusInfoEvent[:43] - rule_3 updated: NOT_INITIALIZED
14:28:16.658 DEBUG o.e.s.c.c.r.AbstractManagedProvider[:62] - Added new element rule_3 to ManagedRuleProvider.
14:28:16.658 INFO  smarthome.event.RuleAddedEvent[:43] - Rule 'rule_3' has been added.

1) The first one NOT_INITIALIZED must be skiped
2) Order of events should be:
- Rule 'rule_3' has been added.
- rule_3 updated: NOT_INITIALIZED
- rule_3 updated: IDLE

### PR DESCRIPTION
**Pull request by danchom**
…der.

Signed-off-by: Yordan Mihaylov <j.mihailov@prosyst.bg>

### CODE REVIEW

File: bundles/automation/org.eclipse.smarthome.automation.core/src/main/java/org/eclipse/smarthome/automation/core/internal/RuleRegistryImpl.java [hunk 1]
```diff
@@ -504,8 +504,8 @@ public void added(Provider<Rule> provider, Rule element) {
             String rUID = ruleEngine.getUniqueId();
             ruleWithUID = initRuleId(rUID, element);
         }
-        super.added(provider, ruleWithUID);
```
Comment (kaikreuzer) [inline bundles/automation/org.eclipse.smarthome.automation.core/src/main/java/org/eclipse/smarthome/automation/core/internal/RuleRegistryImpl.java]: Is it safe to change this order? 
You are now sending the event, before the rule is available in the registry - why is this necessary?

Comment (danchom) [inline bundles/automation/org.eclipse.smarthome.automation.core/src/main/java/org/eclipse/smarthome/automation/core/internal/RuleRegistryImpl.java]: It is not available in registry, but it is already stored in RuleManagedProvider. IMHO added/updated events should be before the state events of rule. Otherwise we can  receive a state events for rules which are not available in the registry (when the add method is failed)

Comment (kaikreuzer) [inline bundles/automation/org.eclipse.smarthome.automation.core/src/main/java/org/eclipse/smarthome/automation/core/internal/RuleRegistryImpl.java]: Looking at the code, I would actually suggest to move the "post event" line to `notifyListenersAboutXXXXElement` - this is where events are sent in ItemRegistryImpl and ThingRegistryImpl, so doing the very same here would imho make sense, wdyt?

Comment (danchom) [inline bundles/automation/org.eclipse.smarthome.automation.core/src/main/java/org/eclipse/smarthome/automation/core/internal/RuleRegistryImpl.java]: It will not be enough, because the status events are fired when the rule is added into RuleEngine, which is done in the onAddElement() method at the moment. The logic should be moved from onXXXElement() to  notifyListenersAboutXXXXElement(). I'll try to fix this.
BTW wdyt about removing redundant NOT_INITIALIZED event.
Up to now we have:
added -> not_initillized -> idle
the proposal is:
added -> idle 
or 
added -> not_initilized.

Comment (kaikreuzer) [inline bundles/automation/org.eclipse.smarthome.automation.core/src/main/java/org/eclipse/smarthome/automation/core/internal/RuleRegistryImpl.java]: > BTW wdyt about removing redundant NOT_INITIALIZED event.

You are referring to [this suggetion](https://github.com/eclipse/smarthome/issues/2823), right? If so, please see my comment on that issue.

**Comment by danchom** [comment]
Implemented:
added -> UNINITIALIZED(without event) -> INITIALIZING -> IDLE
added -> UNINITIALIZED(without event) -> INITIALIZING -> UNINITIALIZED (with details)
updated-> UNINITIALIZED(without event) -> INITIALIZING -> IDLE
updated ->UNINITIALIZED(without event) -> INITIALIZING -> UNINITIALIZED (with details)
enabled -> DISABLED ->UNINITIALIZED(without event) -> INITIALIZING -> IDLE
enabled -> DISABLED ->UNINITIALIZED(without event) -> INITIALIZING ->  UNINITIALIZED (with details)

**Comment by danchom** [comment]
Could you please review this PR.
